mx.sulong/tools.py

```diff
@@ -0,0 +1,176 @@
+ProgrammingLanguage.register('LLVMBC', 'bc')
+ProgrammingLanguage.register('LLVMSU', 'su')
+ProgrammingLanguage.register('EXEC', 'out')
+
+
+class Optimization(object):
+    class Opt(object):
+        def __init__(self, name, flags):
+            self.name = name
+            self.flags = flags
+
+    @staticmethod
+    def register(name, *flags):
+        setattr(Optimization, name, Optimization.Opt(name, list(flags)))
+
+Optimization.register('NONE')
+Optimization.register('BB_VECTORIZE')
+Optimization.register('O1', '-O1')
+Optimization.register('O2', '-O2')
+Optimization.register('O3', '-O3')
+
+
+class Tool(object):
+    def supports(self, language):
+        return language in self.supportedLanguages
+
+class ClangCompiler(Tool):
+    def __init__(self):
+        self.name = 'clang'
+        self.supportedLanguages = [ProgrammingLanguage.C, ProgrammingLanguage.C_PLUS_PLUS, ProgrammingLanguage.OBJECTIVE_C]
+
+    def run(self, inputFile, outputFile, flags):
+        try:
+            f = open('/dev/null', 'w')
+            return mx.run([mx_sulong.findLLVMProgram('clang'), '-S', '-emit-llvm', '-o', outputFile] + flags + [inputFile], out=f, err=f)
+        except SystemExit:
+            print 'Cannot compile %s' % (inputFile)
+        return -1
+
+    def compileReferenceFile(self, inputFile, outputFile, flags):
+        try:
+            f = open('/dev/null', 'w')
+            return mx.run([mx_sulong.findLLVMProgram('clang'), '-o', outputFile] + flags + [inputFile], out=f, err=f)
+        except SystemExit:
+            print 'Cannot compile %s' % (inputFile)
+        return -1
+
+class Opt(Tool):
+    def __init__(self, name, passes):
+        self.name = name
+        self.supportedLanguages = [ProgrammingLanguage.LLVMIR]
+        self.passes = passes
+
+    def supports(self, language):
+        print 'supports? %s' %(language.name)
+        return language in self.supportedLanguages
+
+    def run(self, inputFile, outputFile, flags):
+        return mx.run([mx_sulong.findLLVMProgram('opt')] + ['-S'] + self.passes +  [inputFile] + ['-o', outputFile])
+
+Tool.CLANG = ClangCompiler()
+Tool.BB_VECTORIZE = Opt('BB_VECTORIZE', ['-functionattrs', '-instcombine', '-always-inline', '-jump-threading', '-simplifycfg', '-mem2reg', '-scalarrepl', '-bb-vectorize'])
+
+
+class Runtime(object):
+    def supports(self, language):
+        return language in self.supportedLanguages
+
+class LLIRuntime(Runtime):
+    def __init__(self):
+        self.name = 'lli'
+        self.supportedLanguages = [ProgrammingLanguage.LLVMIR, ProgrammingLanguage.LLVMBC]
+
+    def run(self, f, args, vmArgs=None, out=None):
+        return mx.run([mx_sulong.findLLVMProgram('lli'), f] + args, nonZeroIsFatal=False)
+
+Runtime.LLI = LLIRuntime()
+
+
+def getFileExtension(f):
+    _, ext = os.path.splitext(f)
+    return ext[1:]
+
+def getOutputName(inputFile, outputDir, tool, optimization, target):
+    base, _ = os.path.splitext(inputFile)
```

Comment:
It would be nice if you could add a one-line `"""` comment to explain what these functions do, similar to how we document in the other mx files